### PR TITLE
[core] Merge PmdAnalysis into pmd/7.0.x

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Merge master into pmd/7.0.x, as #3785 and #3692 conflict a lot. I removed most of the deprecated members of `PMD.java` because they're impossible to maintain (processFiles is kept for compatibility with mvn pmd plugin, but it should be updated to use PmdAnalysis after PMD 6.44.0 is out).

Please look at the last commits for what happened after the merge.

Note that this contains #3822.

## Related issues

- Fixes #

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by oowekyala** [comment]
note: the pmd-dist ITs are still failing... will investigate this weekend

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 30 violations,<br />
introduces 22 new violations, 0 new errors and 0 new configuration errors,<br />
removes 30 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/6974d485f8e94e078d612e8859411466/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 57268 violations,<br />
introduces 34612 new violations, 7 new errors and 0 new configuration errors,<br />
removes 138164 violations, 25 errors and 3 configuration errors.<br />
<a href="https://chunk.io/pmd/6974d485f8e94e078d612e8859411466/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: docs/pages/pmd/userdocs/tools/java-api.md [hunk 1]
```diff
@@ -80,66 +80,75 @@ public class PmdExample {
...
     
     ```java
-    try (GlobalAnalysisListener listener = xmlRenderer.newListener()) {
-        PMD.processFiles(configuration, ruleSets, files, listener);
-    } finally {
-        ClassLoader auxiliaryClassLoader = configuration.getClassLoader();
-        if (auxiliaryClassLoader instanceof ClasspathClassLoader) {
-            ((ClasspathClassLoader) auxiliaryClassLoader).close();
-        }
+    try (PmdAnalysis pmd = PmdAnalysis.create(configuration)) {
```
Comment (adangel) [inline docs/pages/pmd/userdocs/tools/java-api.md:136]: That is one example, I was not sure about - but it seems, that GlobalAnalysisListener is not anymore the API to use when calling PMD and is replaced by PmdAnalysis.

File: pmd-core/src/main/java/net/sourceforge/pmd/PmdAnalysis.java [hunk 2]
```diff
@@ -100,49 +110,112 @@ private PmdAnalysis(PMDConfiguration config) {
...
+    /**
+     * Add a new listener. As per the contract of {@link GlobalAnalysisListener},
+     * this object must be ready for interaction. However, nothing will
+     * be done with the listener until {@link #performAnalysis()} is called.
+     * The listener will be closed by {@link #performAnalysis()}, or
+     * {@link #close()}, whichever happens first.
+     *
+     * @throws NullPointerException If the parameter is null
+     */
+    public void addListener(GlobalAnalysisListener listener) {
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/PmdAnalysis.java:206]: ok, so here landed GlobalAnalysisListener - you can add a listener like a renderer...

**Review by adangel — APPROVED** [review]
Thanks for doing the merge!
It was a bit difficult to review - but now I think I understand how GlobalAnalysisListener fits into the picture. At some point, we might add a usage sample in the java api doc.

Btw. this PR is still in draft - is it ready?

**Comment by oowekyala** [comment]
> Btw. this PR is still in draft - is it ready?

It is, I forgot to update it

**Pull request merged**